Refactoring/gcp/cloudsql

Remi05 opened the pull request:
Migrated GCP's Cloud SQL service to the new architecture.

Note that `gcp/resources/projects.py` and `gcp/facade/base.py` are the same as in the [Stackdriver logging PR](https://github.com/nccgroup/ScoutSuite/pull/229) and the [IAM PR](https://github.com/nccgroup/ScoutSuite/pull/230); they are simply shared by all services and will appear in all GCP refactoring PRs until one of the services is merged.

j4v commented:
:clap: :clap: :clap:

Aboisier commented on ScoutSuite/__main__.py:
```diff
@@ -62,17 +61,12 @@
+                                                    mfa_serial=args.get('mfa_serial'),
+                                                    mfa_code=args.get('mfa_code'),
+                                                    user_account=args.get('user_account'),
+                                                    service_account=args.get('service_account'),
```
We keep reverting that change, can we just leave it like it was haha

misg commented:
Could you please merge with `refactoring/gcp-resource-configs` to avoid the 263 commits and 86 files changed (I can't do my review in those conditions, sorry :sob:).

Aboisier commented on ScoutSuite/providers/gcp/facade/utils.py:
```diff
@@ -0,0 +1,19 @@
+        await run_concurrently(
+            lambda: GCPFacadeUtils._get_all(resources, resource_key, request, resources_group)
+        )
+        return resources
```
I'm nitpicking, but I think it makes more sense to put `get_all` before `_get_all` (when you read the file, you start from a high level and you go deeper, it makes it more readable IMO).

Aboisier commented on ScoutSuite/providers/gcp/facade/utils.py:
```diff
@@ -0,0 +1,19 @@
+import httplib2
```
Unused import

Aboisier commented on ScoutSuite/providers/gcp/facade/utils.py:
```diff
@@ -0,0 +1,19 @@
+import httplib2
+import googleapiclient
```
Unused import

Aboisier commented on ScoutSuite/providers/gcp/facade/utils.py:
```diff
@@ -0,0 +1,19 @@
+
+class GCPFacadeUtils:
+    @staticmethod
+    def _get_all(resources, resource_key, request, resources_group):
```
Are there  parameters you can add type hints to?

Remi05 replied:
I didn't find any types for `resources`, `request` and `resources_group` as the library is auto-generated using the Google Cloud API description.

Aboisier commented on ScoutSuite/providers/gcp/facade/utils.py:
```diff
@@ -0,0 +1,19 @@
+            request = resources_group.list_next(previous_request=request, previous_response=response)
+
+    @staticmethod
+    async def get_all(resource_key, request, resources_group):
```
Same here for the type hints

Remi05 commented:
Fixed UI and added left menu for projects.

![cloudsql_projects](https://user-images.githubusercontent.com/12770956/55366679-be4d5c00-54b7-11e9-842f-73a3e4867163.png)

Aboisier commented on ScoutSuite/providers/gcp/facade/cloudsql.py:
```diff
@@ -0,0 +1,27 @@
+        cloudsql_client = self._get_client()
+        backups_group = cloudsql_client.backupRuns()
+        request = backups_group.list(project=project_id, instance=instance_name)
+        return await GCPFacadeUtils.get_all('items', request, backups_group)
```
Shouldn't this be parallelized?

misg replied:
It's already parallelized in the `get_all` method.

Aboisier replied:
Oh my bad, I looked at the `_get_all`.

Aboisier commented on ScoutSuite/providers/gcp/facade/cloudsql.py:
```diff
@@ -0,0 +1,27 @@
+        cloudsql_client = self._get_client()
+        instances_group = cloudsql_client.instances()
+        request = instances_group.list(project=project_id)
+        return await GCPFacadeUtils.get_all('items', request, instances_group)
```
Same here, shouldn't this be parallelized?

Aboisier commented on ScoutSuite/providers/gcp/resources/cloudsql/backups.py:
```diff
@@ -0,0 +1,24 @@
+    async def fetch_all(self):
+        raw_backups = await self.gcp_facade.cloudsql.get_backups(self.project_id, self.instance_name)
+        for raw_backup in raw_backups:
+            if raw_backup['status'] == 'SUCCESSFUL':
```
I guess it was like that before, but what about the failed backups?

Remi05 replied:
I guess they are considered not to be relevant, but @j4v would know more!

Aboisier commented on ScoutSuite/providers/gcp/resources/cloudsql/database_instances.py:
```diff
@@ -0,0 +1,49 @@
+        for raw_instance in raw_instances:
+            instance_id, instance = self._parse_instance(raw_instance)
+            self[instance_id] = instance
+            await self._fetch_children(self[instance_id], gcp_facade = self.gcp_facade, project_id = self.project_id, instance_name = instance['name'])
```
I believe this should be parallelized using `asyncio.ensure_future`.

Remi05 replied:
Done!
